vsphere/check.py

```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
+                    for datastore in c.datastore:
+                        datastores.append(datastore.name)   # datastore._moId - GUID of store
+                    topology_tags["topo_type"] = VSPHERE_COMPONENT_TYPE.DATACENTER
+                    topology_tags["datastores"] = datastores
+                    topology_tags["name"] = c.name
+                    topology_tags["id"] = c._moId
+                    topology_tags["layer"] = TOPOLOGY_LAYERS.DATACENTER
+                    topology_tags["domain"] = domain
+
+                    computeresources = []
+                    clustercomputeresources = []
+
+                    for computeres in c.hostFolder.childEntity:
+                        if isinstance(computeres, vim.ComputeResource):
+                            computeresources.append(computeres.name)
+                        elif isinstance(computeres, vim.CloudComputeResource):
+                            clustercomputeresources.append(computeres.name)
+
+                    topology_tags["computeresources"] = computeresources
+                    topology_tags["clustercomputeresources"] = clustercomputeresources
+
+                    add_label_pair(labels, "name", topology_tags["name"])
+
+                    hostname = None
+                topology_tags["labels"] = labels
+                obj_list.append(dict(mor_type=vimtype, mor=c, hostname=hostname, topo_tags = topology_tags))
+
+        return obj_list
+
+    def get_topologyitems_sync(self, instance, tags=[], regexes=None, include_only_marked=False):
+        server_instance = self._get_server_instance(instance)
+        content = server_instance.RetrieveContent()
+        domain = instance["host"]  # candidate also name
+
+        vms = self._vsphere_objs(content, "vm", domain)
+        hosts = self._vsphere_objs(content, "host", domain)
+        datacenters = self._vsphere_objs(content, "datacenter", domain)
+        datastores = self._vsphere_objs(content, "datastore", domain)
+        clustercomputeresource = self._vsphere_objs(content, "clustercomputeresource", domain)
+        computeresource = self._vsphere_objs(content, "computeresource", domain)
+
+        return {
+            "vms": vms,
+            "hosts": hosts,
+            "datacenters": datacenters,
+            "datastores": datastores,
+            "clustercomputeresource": clustercomputeresource,
+            "computeresource": computeresource
+        }
+
+    def collect_topology(self,instance):
+
+        def build_id(vsphere_url, object_type, object_name):
+            return "urn://vsphere/{0}/{1}/{2}".format(vsphere_url, object_type, object_name)
+
+        def build_type(object_type):
+            return {"name": object_type}
+
+        topology_items = self.get_topologyitems_sync(instance)
+        vsphere_url = instance.get("host")
+        instance_key = {"type": self.INSTANCE_TYPE, "url": vsphere_url}
+
+        self.start_snapshot(instance_key)
+
+        for vm in topology_items["vms"]:
```

Comment:
I see a pattern here; in `_vsphere_objs` there is an `if`-statement determining the type and here we do some actions based on the type as well, should we merge that together? Maybe a method per type to handle everything for that type?